src/mito2/src/memtable/merge_tree/data.rs

```diff
@@ -519,26 +522,137 @@ impl<'a> DataPartEncoder<'a> {
+            .field("current_pk_index", &self.current_pk_index)
+            .finish()
+    }
+}
+
+impl DataPartReader {
+    pub fn new(data: Bytes, batch_size: Option<usize>) -> Result<Self> {
+        let mut builder =
+            ParquetRecordBatchReaderBuilder::try_new(data).context(error::ReadDataPartSnafu)?;
+        if let Some(batch_size) = batch_size {
+            builder = builder.with_batch_size(batch_size);
+        }
+        let parquet_reader = builder.build().context(error::ReadDataPartSnafu)?;
+        let mut reader = Self {
+            inner: parquet_reader,
+            current_pk_index: None,
+            current_range: 0..0,
+            current_batch: None,
+        };
+        reader.next()?;
+        Ok(reader)
+    }
+
+    /// Returns false if current reader is exhausted.
+    pub(crate) fn is_valid(&self) -> bool {
+        self.current_pk_index.is_some()
+    }
+
+    /// Returns current pk index.
+    ///
+    /// # Panics
+    /// If reader is exhausted.
+    pub(crate) fn current_pk_index(&self) -> PkIndex {
+        self.current_pk_index.expect("DataPartReader is exhausted")
+    }
+
+    /// Returns current data batch of reader.
+    /// # Panics
+    /// If reader is exhausted.
+    pub(crate) fn current_data_batch(&self) -> DataBatch {
+        let rb = self.current_batch.as_ref().unwrap();
+        let pk_index = self.current_pk_index.unwrap();
+        let range = self.current_range.clone();
+        DataBatch {
+            pk_index,
+            rb,
+            range,
+        }
+    }
+
+    pub(crate) fn next(&mut self) -> Result<()> {
+        if let Some((next_pk, range)) = self.search_next_pk_range() {
+            // first try to search next pk in current record batch.
+            self.current_pk_index = Some(next_pk);
+            self.current_range = range;
+        } else {
+            // current record batch reaches eof, fetch next record batch from parquet reader.
+            if let Some(rb) = self.inner.next() {
+                let rb = rb.context(error::ComputeArrowSnafu)?;
+                self.current_range = 0..0;
+                self.current_batch = Some(rb);
+                return self.next();
+            } else {
+                // parquet is also exhausted
+                self.current_pk_index = None;
+                self.current_batch = None;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Searches next primary key along with it's offset range inside record batch.
+    fn search_next_pk_range(&self) -> Option<(PkIndex, Range<usize>)> {
+        self.current_batch.as_ref().and_then(|b| {
+            // safety: PK_INDEX_COLUMN_NAME must present in record batch yielded by data part.
+            let pk_array = pk_index_array(b);
+            search_next_pk_range(pk_array, self.current_range.end)
+        })
+    }
+}
+
+/// Parquet-encoded `DataPart`.
+pub struct ParquetPart {
+    data: Bytes,
+    metadata: FileMetaData,
```

Comment:
Do we use the metadata?

Reply:
> Do we use the metadata?

ts range for each part is useful for skip merge process. But I'm not sure what struct to use. Simply put the parquet metadata as a placeholder.